File: ch13/Chapter13-CS/OurFirstWebApi-VB/Controllers/ProductsController.vb
Imports System.Data.Entity
Imports System.Data.Entity.Infrastructure
Imports System.Net
Imports System.Net.Http
Imports System.Web.Http
Namespace OurFirstWebApi_VB
    Public Class ProductsController
        Inherits System.Web.Http.ApiController

        Private db As New AdventureWorksEntities

        ' GET api/Products
        Function GetProducts() As IEnumerable(Of Product)
            Dim products = db.Products.Include(Function(p) p.ProductCategory).Include(Function(p) p.ProductModel)
            Return products.AsEnumerable()
        End Function

        ' GET api/Products/5
        Function GetProduct(ByVal id As Integer) As Product
            Dim product As Product = db.Products.Find(id)
            If IsNothing(product) Then
                Throw New HttpResponseException(Request.CreateResponse(HttpStatusCode.NotFound))
            End If
            Return product
        End Function

        ' PUT api/Products/5
        Function PutProduct(ByVal id As Integer, ByVal product As Product) As HttpResponseMessage
            If ModelState.IsValid And id = product.ProductID Then
                db.Entry(product).State = EntityState.Modified

                Try
                    db.SaveChanges()
                Catch ex As DbUpdateConcurrencyException
                    Return Request.CreateResponse(HttpStatusCode.NotFound)
                End Try

                Return Request.CreateResponse(HttpStatusCode.OK)
            Else
                Return Request.CreateResponse(HttpStatusCode.BadRequest)
            End If
        End Function

        ' POST api/Products
        Function PostProduct(ByVal product As Product) As HttpResponseMessage
            If ModelState.IsValid Then
                db.Products.Add(product)
                db.SaveChanges()

                Dim response As HttpResponseMessage = Request.CreateResponse(HttpStatusCode.Created, product)
                response.Headers.Location = New Uri(Url.Link("DefaultApi", New With {.id = product.ProductID}))
                Return response
            Else
                Return Request.CreateResponse(HttpStatusCode.BadRequest)
            End If
        End Function

        ' DELETE api/Products/5
        Function DeleteProduct(ByVal id As Integer) As HttpResponseMessage
            Dim product As Product = db.Products.Find(id)
            If IsNothing(product) Then
                Return Request.CreateResponse(HttpStatusCode.NotFound)
            End If

            db.Products.Remove(product)

            Try
                db.SaveChanges()
            Catch ex As Exception
                Return Request.CreateResponse(HttpStatusCode.NotFound)
            End Try

            Return Request.CreateResponse(HttpStatusCode.OK, product)
        End Function

        Protected Overrides Sub Dispose(ByVal disposing As Boolean)
            db.Dispose()
            MyBase.Dispose(disposing)
        End Sub
    End Class
End Namespace

File: ch13/Chapter13-CS/OurFirstWebApi-CS/Controllers/ProductsController.cs
﻿using System;
using System.Collections.Generic;
using System.Data;
using System.Data.Entity;
using System.Data.Entity.Infrastructure;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Web;
using System.Web.Http;
using OurFirstWebApi_CS.Models;

namespace OurFirstWebApi_CS.Controllers
{
public class ProductsController : ApiController
{
    private AdventureWorksEntities db = new AdventureWorksEntities();

    public ProductsController()
    {
        db.Configuration.ProxyCreationEnabled = false;
    }

    // GET api/Products
    public IEnumerable<Product> GetProducts()
    {
        var products = db.Products.Include(p => p.ProductCategory).Include(p => p.ProductModel);
        return products.AsEnumerable();
    }

    // GET api/Products/5
    public Product GetProduct(int id)
    {
        Product product = db.Products.Find(id);
        if (product == null)
        {
            throw new HttpResponseException(Request.CreateResponse(HttpStatusCode.NotFound));
        }

        return product;
    }

    // PUT api/Products/5
    public HttpResponseMessage PutProduct(int id, Product product)
    {
        if (ModelState.IsValid && id == product.ProductID)
        {
            db.Entry(product).State = EntityState.Modified;

            try
            {
                db.SaveChanges();
            }
            catch (DbUpdateConcurrencyException)
            {
                return Request.CreateResponse(HttpStatusCode.NotFound);
            }

            return Request.CreateResponse(HttpStatusCode.OK);
        }
        else
        {
            return Request.CreateResponse(HttpStatusCode.BadRequest);
        }
    }

    // POST api/Products
    public HttpResponseMessage PostProduct(Product product)
    {
        if (ModelState.IsValid)
        {
            db.Products.Add(product);
            db.SaveChanges();

            HttpResponseMessage response = Request.CreateResponse(HttpStatusCode.Created, product);
            response.Headers.Location = new Uri(Url.Link("DefaultApi", new { id = product.ProductID }));
            return response;
        }
        else
        {
            return Request.CreateResponse(HttpStatusCode.BadRequest);
        }
    }

    // DELETE api/Products/5
    public HttpResponseMessage DeleteProduct(int id)
    {
        Product product = db.Products.Find(id);
        if (product == null)
        {
            return Request.CreateResponse(HttpStatusCode.NotFound);
        }

        db.Products.Remove(product);

        try
        {
            db.SaveChanges();
        }
        catch (DbUpdateConcurrencyException)
        {
            return Request.CreateResponse(HttpStatusCode.NotFound);
        }

        return Request.CreateResponse(HttpStatusCode.OK, product);
    }

    protected override void Dispose(bool disposing)
    {
        db.Dispose();
        base.Dispose(disposing);
    }
}
}